### ISSUE TITLE
Bike routing penalty for flyover/overpass bridge

### ISSUE THREAD

**Issue by nemobis** [graphhopper/graphhopper#597]
I find GraphHopper almost magic for bike routing and I'm recommending it to everyone since a few months, thanks a lot.

I've been fighting a bit with the directions at https://graphhopper.com/maps/?point=45.511279%2C9.254265&point=45.517805%2C9.266946&vehicle=bike&weighting=fastest&elevation=true&layer=TF+Cycle , IMHO blatantly wrong because they ignore a provincial cycling route (via Idro on naviglio Martesana) and suggest instead to go in the middle of a motorway junction (via Padova at Cascina Gobba, with a cavalcavia). The desired routing is: https://graphhopper.com/maps/?point=45.511279%2C9.254265&point=45.517805%2C9.266946&vehicle=bike&weighting=fastest&elevation=true&layer=TF+Cycle

![gobba-cavalcavia](https://cloud.githubusercontent.com/assets/901528/11639499/89a88154-9d2d-11e5-9895-3d2d6743d7f1.png)

![gobba-bici](https://cloud.githubusercontent.com/assets/901528/11639503/8db008b2-9d2d-11e5-829e-a8cee69606e3.png)

Multiple possible errors or deficiencies are discussed at https://lists.openstreetmap.org/pipermail/talk-it/2015-December/thread.html#50340 ; I hope the matter is not with the alleged 2 meter saving in elevation or with the alleged 90 meters in distance saved.

One thing I think should be considered is that the bridge, especially being over a motorway, is moderately steep; currently bridges seem to have no weight in that sense: https://github.com/graphhopper/graphhopper/blob/fed03547acea0b0edbbb90dc723b944250463ef9/core/src/main/java/com/graphhopper/routing/util/Bike2WeightFlagEncoder.java#L161
I think you might want to estimate a steepness, or a slowdown, depending factors such as length of the bridge and kind of underlying obstacle.

As far as I can see, the data is correct: https://wiki.openstreetmap.org/wiki/Key:ele , https://wiki.openstreetmap.org/wiki/Key:bridge , https://wiki.openstreetmap.org/wiki/Key:height don't mention any further detail on bridge steepness being recommended or even allowed ; I asked at https://wiki.openstreetmap.org/wiki/Talk:Key:height#Bridge_steepness just in case.

**Comment by ratrun** [comment]
Nice to hear that you find graphhopper bike-routing useful. Please note that the observed example does not have anything to do with elevation weighting. One can verify this easily by switching from the default bike vehicle to race bike. The race bike routing currently does not take elevation data into regard. In your case you get the same unwanted result, which proofs that it is not a problem of bad elevation data.

By setting a "via" point 

https://graphhopper.com/maps/?point=45.511279%2C9.254265&point=45.518745%2C9.259876&point=45.517805%2C9.266946&locale=de-DE&vehicle=racingbike&weighting=fastest&elevation=true&layer=TF%20Cycle

one can see that the desired route is slightly longer (1.7km in comparison to 1.61km). 

In a more detailed investigation I discovered the reason: Via Idro is tagged with a maxspeed of just 15 km/h. I verified this by removing the `maxspeed` tag on all three Via Idro segments. In this case the desired route is found. 
I have started investigating the maxspeed handling and discovered that we are using an inherited car specific applyMaxSpeed from the AbstractFlagEncoder class, which does not make any sense for bicycles.

**Comment by nemobis** [comment]
ratrun, 08/12/2015 10:04:
> In your case you get the same unwanted
> result, which proofs that it is not a problem of bad elevation data.

Thanks for confirming. In my opinion, elevation over bridges should be 
considered in some way, although it's not the main factor in the 
specific example.

> In a more detailed investigation I discovered the reason: Via Idro is
> tagged with a maxspeed of just 15 km/h. I verified this by removing the
> |maxspeed| tag on all three Via Idro segments. In this case the desired
> route is found.
> I have started investigating the maxspeed handling and discovered that
> we are using an inherited car specific applyMaxSpeed from the
> AbstractFlagEncoder class, which does not make any sense for bicycles.

Great find! My proposal 
https://github.com/graphhopper/graphhopper/issues/598 goes even further.

**Comment by karussell** [comment]
Thanks @nemobis for the kind words, really a community effort :) !

Also as always thanks to @ratrun for the investigation!

> I have started investigating the maxspeed handling and discovered that we are using an inherited car specific applyMaxSpeed from the AbstractFlagEncoder class, which does not make any sense for bicycles.

This sounds ugly as I thought I included a unit test some time ago for exactly this, hmmh. Will have a look.

**Comment by ratrun** [comment]
@nemobis: Maybe you are right that we could put some penalty on bridges because of elevation. But currently we can safely neglect such details, because the inaccurate heuristics with SRTM elevation data handling creates much more problems. I believe that it would help if we would take the 'incline' into regard and tune our SRTM elevation weighting results by comparing the computed inclince to the tagged incline value. If the ramps for bridges are tagged with an 'incline', then this would better cover the bridge situation.
But fot the implementation of 'incline' handling I fear that there are lots of segments taged with a wrong inclince value, which does not match to the direction of the way, similar as the problems of wrong direction of flowing water streams.
@karussell : It would be great if you could comment and confirm or disprove my findings by looking on the tests and the results of 'BikeFlagEncoderTest:testMaxSpeed(). I stepped through with the debugger and discovered the calculations in applyMaxSpeed with 0.9. It would be great to comment on what the simple applyMaxSpeed is supposed to do exactly.

**Comment by karussell** [comment]
> But fot the implementation of 'incline' handling I fear that there are lots of segments taged with a wrong inclince value

That is exactly my feeling. The sources for elevation data should not mix - at least not now to keep it simple.

> It would be great if you could comment and confirm or disprove my findings by looking on the tests

Will do when I find more time - hopefully soonish ;)

> It would be great to comment on what the simple applyMaxSpeed is supposed to do exactly.

It should be better documented ... if the maxspeed tag is available then normally it decreases the speed guessed from the road type (if maxspeed < speed). BUT it can increase this guessed value if force is true, which should be the case for car only.

**Comment by ratrun** [comment]
@karussell: Thank you for the explanation of 'applyMaxSpeed'. I think that I now understood at least one part of the problem. Our speed would be 18km/h, we run applyMaxSpeed with 'false' and get back an even lower speed: 15km/h*0.9 = 14km/h. It does not make much sense to decrease such a low speed for bicycles even further. Most riders will ride faster compared to such low maxspeeds. And in the example, we have ordinary residentials without low maxspeeds as alternatives, which consequently get preferred.

**Comment by ratrun** [comment]
This finding encoveres two separated sub-problems. 
1. Sub-problem 1 is the previously described handling for maxspeed in applyMaxSpeed. #600 fixes this sub-problem.

2. Sub-problem 2 is a too low priority boosting difference. The desired route here is running on a "rcn" cycle route, which gets classified as VERY_NICE. The boosting of such a classification in comparison to mainly tertiary and ordingary cycle ways does not decrease the weighting enough such that we tolerate 90 meters more distance. The solution for this sub-problem is simple in code, but as breaks a lot of tests I haven't fixed it yet because we should agree here about the need for more boosting based on the priorities first. I believe that both theses two sub-problems can be discussed independently. Even tough these two sub-problems are seperated, for simplicity reasons I plan to create a PR for this sub-problem after #600 is decided.

**Comment by karussell** [comment]
I think this issue is fixed now?

**Comment by nemobis** [comment]
The original example is fine now, so we can close. There were multiple factors, so I'm not sure how the routing performs in other cases of motorway bridges, but I'll file a separate task if I find out more.

**Comment by karussell** [comment]
Thanks!
